File: 99-Continue/161/LCA.java
import java.util.*;

public class LCA {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        sc.nextLine();

        String s = sc.nextLine();
        String[] arr = s.split(" ");

        BinaryTree tree = new BinaryTree();
        tree.buildTree(arr);

        String s2 = sc.nextLine();
        String[] sarr = s2.split(" ");

        System.out.println(tree.leastCommanAncestor(sarr));

        sc.close();
    }
}

class Node {
    int data;
    Node left, right;

    Node(int data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BinaryTree {

    Node root;

    BinaryTree() {
        root = null;
    }

    Node buildTree(String[] arr) {

        if (arr.length == 0 || arr[0].equals("null"))
            return null;

        root = new Node(Integer.parseInt(arr[0]));

        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);

        int i = 1;

        while (!queue.isEmpty() && i < arr.length) {

            Node current = queue.poll();

            if (i < arr.length && !arr[i].equals("null")) {
                current.left = new Node(Integer.parseInt(arr[i]));
                queue.offer(current.left);
            }
            i++;

            if (i < arr.length && !arr[i].equals("null")) {
                current.right = new Node(Integer.parseInt(arr[i]));
                queue.offer(current.right);
            }
            i++;
        }

        return root;
    }

    int leastCommanAncestor(String[] arr) {

        int a = Integer.parseInt(arr[0]);
        int b = Integer.parseInt(arr[1]);

        Node ans = lca(root, a, b);

        return ans == null ? -1 : ans.data;
    }

    Node lca(Node root, int a, int b) {

        if (root == null)
            return null;

        if (root.data == a || root.data == b)
            return root;

        Node left = lca(root.left, a, b);
        Node right = lca(root.right, a, b);

        if (left != null && right != null)
            return root;

        return (left != null) ? left : right;
    }
}

File: 99-Continue/161/lca.py
from collections import deque

class Node:
    def __init__(self,data:int):
        self.left = self.right = None
        self.data = data

class BinaryTree:
    def __init__(self):
        self.root = None
        
    def buildTree(self,arr:list[str])->Node:
        if len(arr)==0:
            return self.root
    
        self.root = Node(int(arr[0]))
        queue = deque([self.root])
        i=1
        while queue and i<len(arr):
            current = queue.popleft()
            if i<len(arr) and arr[i]!="null":
                current.left = Node(int(arr[i]))
                queue.append(current.left)
            i+=1
            if i<len(arr) and arr[i]!="null":
                current.right = Node(int(arr[i]))
                queue.append(current.right)
            i+=1
        return self.root
    
    def lca(self,root:Node,a:int,b:int) -> Node:
        if root is None:
            return root
        
        if root.data==a or root.data==b:
            return root
        
        left = self.lca(root.left,a,b)
        right = self.lca(root.right,a,b)
        
        if left is not None and right is not None:
            return root
        
        return left if left is not None else right
        
def main():
    n = int(input())
    tree = BinaryTree()
    lsr = input().split()
    tree.buildTree(lsr)
    insr = list(map(int,input().split()))
    print(tree.lca(tree.root,insr[0],insr[1]).data)

if __name__ == "__main__":
    main()
    